base/abstractarray.jl

```diff
@@ -1048,6 +1048,50 @@ function _setindex!(::IndexCartesian, A::AbstractArray, v, I::Vararg{Int,M}) whe
     r
 end
 
+## rudimentary aliasing detection ##
+"""
+    unalias(dest, A)
+
+Return either A or a copy of A, in a rough effort to prevent modifications to `dest` from
+affecting the returned object. No guarantees are provided, and each custom array must
+opt-into aliasing detection by overloading this method by specializing on the second argument.
+
+This function must return an object of exactly the same type as A for performance and type-stability.
+
+See also `mightalias` and `dataids`.
+"""
+unalias(dest, A) = A
+unalias(dest, A::Array) = mightalias(dest, A) ? copy(A) : A
+@inline unalias(dest, As::Tuple) = (unalias(dest, As[1]), unalias(dest, tail(As))...)
+unalias(dest, As::Tuple{}) = ()
+
+"""
+    mightalias(A::AbstractArray, B::AbstractArray)
+
+Perform a rudimentary test to check if arrays A and B might share the same memory.
+
+Defaults to false unless `dataids` is specialized for both `A` and `B`.
```

Comment:
When in doubt, wouldn't it be better to return `true`? More generally, shouldn't the functions require/guarantee that any aliasing is detected and that in doubt a copy will be made? The use of terms like "rough" and "rudimentary" doesn't inspire confidence. ;-) We don't want silent corruption to happen by default...

Or it frequently impossible to determine whether arrays alias?

Reply:
Indeed, I don't want to inspire confidence.  That's why this isn't exported.  I just want to do incrementally better than we had been doing before for the builtin array types.  There's an asymmetry in costs here — returning `true` by default would lead to _lots_ of unnecessary copies.  But we also need to perform this check quite frequently, so I don't want the check itself to be expensive.  This is the tradeoff I came to.

I initially had default fallbacks for strided arrays based on `pointer(A, firstindex(A))` and `pointer(A, lastindex(A))` and all other arrays based on `objectid(A)` — that led to #25858.  We could try to extend this more generically later.